### 5) . 1

In [9]:
import pandas as pd
fileName = 'zooData.csv'
data = pd.read_csv(fileName)
#print(data);

### 5) .2

In [10]:
data.drop(['animalName'], axis=1, inplace=True)
Y = data['type']
print(Y)

0            mammal
1            mammal
2              fish
3            mammal
4            mammal
5            mammal
6            mammal
7              fish
8              fish
9            mammal
10           mammal
11             bird
12             fish
13     invertebrate
14     invertebrate
15     invertebrate
16             bird
17           mammal
18             fish
19           mammal
20             bird
21             bird
22           mammal
23             bird
24           insect
25        amphibian
26        amphibian
27           mammal
28           mammal
29           mammal
           ...     
71             bird
72     invertebrate
73             fish
74           mammal
75           mammal
76          reptile
77     invertebrate
78             bird
79             bird
80          reptile
81     invertebrate
82             fish
83             bird
84           mammal
85     invertebrate
86             fish
87             bird
88           insect
89        amphibian


In [11]:
data1 =data
data1.drop(['type'], axis=1, inplace=True)
X = data1
print(X)

      hair  feathers   eggs   milk  airborne  aquatic  predator  toothed  \
0     True     False  False   True     False    False      True     True   
1     True     False  False   True     False    False     False     True   
2    False     False   True  False     False     True      True     True   
3     True     False  False   True     False    False      True     True   
4     True     False  False   True     False    False      True     True   
5     True     False  False   True     False    False     False     True   
6     True     False  False   True     False    False     False     True   
7    False     False   True  False     False     True     False     True   
8    False     False   True  False     False     True      True     True   
9     True     False  False   True     False    False     False     True   
10    True     False  False   True     False    False      True     True   
11   False      True   True  False      True    False     False    False   
12   False  

In [12]:
# using training set
from sklearn import tree
clf = tree.DecisionTreeClassifier()
clf = clf.fit(X,Y)
print('Training Accuracy: ',clf.score(X,Y)) # Training accuracy

Training Accuracy:  1.0


In [13]:
# by splitting the data as 2/3 for training and 1/3 for testing
from sklearn.model_selection import train_test_split
X_train , X_test , Y_train , Y_test = train_test_split(X, Y, test_size =0.333,random_state =0)
clf.fit(X_train ,Y_train) # clf is a classifier.
print('Test Accuracy: ',clf.score(X_test ,Y_test )) # Test accuracy

Test Accuracy:  0.8823529411764706


In [15]:
# 10-fold cross validation
from sklearn.model_selection import cross_val_score
scores = cross_val_score(clf, X, Y, cv=10) # 10-fold cross validation
print(scores) # Results for all the folds
print("10CV Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(),scores.std()*2))

[1.         0.91666667 0.91666667 0.90909091 0.9        0.88888889
 1.         1.         1.         1.        ]
10CV Accuracy: 0.95 (+/- 0.09)


C:\Users\user\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:605: Warning: The least populated class in y has only 4 members, which is too few. The minimum number of members in any class cannot be less than n_splits=10.
  % (min_groups, self.n_splits)), Warning)


#### 10 fold cross validation provides more realistic future performance as it uses the whole test set as trainnig set and also as the test set in each fold and another issue is that training set may be irralavent in future with changing environment.Therefore even the traning set have the highest   accuracy it may be useless for future predictions.

### 5) . 3

#####  DecisionTreeClassifier

In [16]:
# by splitting the data as 2/3 for training and 1/3 for testing
from sklearn.metrics import confusion_matrix
from sklearn import tree
clf = tree.DecisionTreeClassifier()
X_train , X_test , Y_train , Y_test = train_test_split(X, Y, test_size =0.333,random_state =0)
clf.fit(X_train ,Y_train) # clf is a classifier.
Y_pred = clf.fit(X_train , Y_train ).predict(X_test)
print(confusion_matrix(Y_test , Y_pred ))

[[ 1  0  0  0  0  0  0]
 [ 0  6  0  0  0  0  0]
 [ 0  0  7  0  0  0  0]
 [ 0  0  0  2  0  0  0]
 [ 0  0  0  0  2  0  0]
 [ 0  0  0  0  0 12  0]
 [ 4  0  0  0  0  0  0]]


##### obviously accuracy is not 100% .because  both 1 & 7 classified as 1

##### MultinomialNB classifier

In [17]:
from sklearn.naive_bayes import MultinomialNB
clf=MultinomialNB() # clf is a classifier.
clf.fit(X,Y)
X_train , X_test , Y_train , Y_test = train_test_split(X, Y, test_size =0.333,random_state =0)
clf.fit(X_train ,Y_train) # clf is a classifier.
Y_pred = clf.fit(X_train , Y_train ).predict(X_test)
print(confusion_matrix(Y_test , Y_pred ))

[[ 0  0  0  0  1  0  0]
 [ 0  6  0  0  0  0  0]
 [ 0  0  7  0  0  0  0]
 [ 0  0  0  2  0  0  0]
 [ 0  0  0  0  2  0  0]
 [ 0  0  0  0  0 12  0]
 [ 0  0  3  0  0  1  0]]


#####  3 errors in this classifier. 1 is predict as 5 , 3  7's  is predict as 3 and one is predict as 6.

##### GaussianNB classifier

In [18]:
from sklearn.naive_bayes import GaussianNB
clf = GaussianNB()
clf.fit(X,Y)
X_train , X_test , Y_train , Y_test = train_test_split(X, Y, test_size =0.333,random_state =0)
clf.fit(X_train ,Y_train) # clf is a classifier.
Y_pred = clf.fit(X_train , Y_train ).predict(X_test)
print(confusion_matrix(Y_test , Y_pred ))

[[ 1  0  0  0  0  0  0]
 [ 0  6  0  0  0  0  0]
 [ 0  0  7  0  0  0  0]
 [ 0  0  0  2  0  0  0]
 [ 0  0  0  0  2  0  0]
 [ 0  0  0  0  0 12  0]
 [ 1  0  0  0  2  1  0]]


##### in above confusion matrix 4 classification errors have happend. (7 classified as 1, 5 & 6)

##### Support Vector Machine

In [19]:
from sklearn import svm
clf = svm.SVC(kernel='linear', C=1,gamma=1).fit(X, Y) # clf is a classifier
clf.fit(X,Y) # Model
X_train , X_test , Y_train , Y_test = train_test_split(X, Y, test_size =0.333,random_state =0)
clf.fit(X_train ,Y_train) # clf is a classifier.
Y_pred = clf.fit(X_train , Y_train ).predict(X_test)
print(confusion_matrix(Y_test , Y_pred ))

[[ 1  0  0  0  0  0  0]
 [ 0  6  0  0  0  0  0]
 [ 0  0  7  0  0  0  0]
 [ 0  0  0  2  0  0  0]
 [ 0  0  0  0  2  0  0]
 [ 0  0  0  0  0 12  0]
 [ 1  0  3  0  0  0  0]]


##### in above confusion matrix 4 classification errors have happend. (7 classified as 1 & 3)

### 5 ) . 4

In [20]:
from sklearn import tree
clf = tree.DecisionTreeClassifier()
clf.fit(X,Y)
from sklearn.model_selection import cross_val_score
scores = cross_val_score(clf, X, Y, cv=10) # 10-fold cross validation
print(scores) # Results for all the folds
print("10CV Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(),scores.std()*2))

[1.         0.91666667 0.91666667 0.90909091 0.8        0.88888889
 1.         1.         1.         1.        ]
10CV Accuracy: 0.94 (+/- 0.13)


C:\Users\user\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:605: Warning: The least populated class in y has only 4 members, which is too few. The minimum number of members in any class cannot be less than n_splits=10.
  % (min_groups, self.n_splits)), Warning)


In [21]:
from sklearn.naive_bayes import GaussianNB
clf = GaussianNB()
clf.fit(X,Y)
from sklearn.model_selection import cross_val_score
scores = cross_val_score(clf, X, Y, cv=10) # 10-fold cross validation
print(scores) # Results for all the folds
print("10CV Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(),scores.std()*2))

[1.         0.91666667 0.91666667 0.90909091 1.         0.88888889
 0.88888889 1.         1.         1.        ]
10CV Accuracy: 0.95 (+/- 0.10)


C:\Users\user\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:605: Warning: The least populated class in y has only 4 members, which is too few. The minimum number of members in any class cannot be less than n_splits=10.
  % (min_groups, self.n_splits)), Warning)


In [22]:
from sklearn.naive_bayes import MultinomialNB
clf=MultinomialNB() # clf is a classifier.
clf.fit(X,Y)
from sklearn.model_selection import cross_val_score
scores = cross_val_score(clf, X, Y, cv=10) # 10-fold cross validation
print(scores) # Results for all the folds
print("10CV Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(),scores.std()*2))

[0.92307692 0.91666667 0.83333333 0.90909091 0.9        0.88888889
 0.88888889 0.77777778 1.         0.875     ]
10CV Accuracy: 0.89 (+/- 0.11)


C:\Users\user\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:605: Warning: The least populated class in y has only 4 members, which is too few. The minimum number of members in any class cannot be less than n_splits=10.
  % (min_groups, self.n_splits)), Warning)


In [23]:
from sklearn import neighbors , datasets
clf = neighbors.KNeighborsClassifier(n_neighbors =1) # 1-Nearest Neighbor
clf.fit(X,Y) # Model
from sklearn.model_selection import cross_val_score
scores = cross_val_score(clf, X, Y, cv=10) # 10-fold cross validation
print(scores) # Results for all the folds
print("10CV Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(),scores.std()*2))

[1.         0.91666667 1.         1.         0.9        1.
 1.         1.         1.         1.        ]
10CV Accuracy: 0.98 (+/- 0.07)


C:\Users\user\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:605: Warning: The least populated class in y has only 4 members, which is too few. The minimum number of members in any class cannot be less than n_splits=10.
  % (min_groups, self.n_splits)), Warning)


In [24]:
from sklearn import svm
clf = svm.SVC(kernel='linear', C=1,gamma=1).fit(X, Y) # clf is a classifier
clf.fit(X,Y) # Model
from sklearn.model_selection import cross_val_score
scores = cross_val_score(clf, X, Y, cv=10) # 10-fold cross validation
print(scores) # Results for all the folds
print("10CV Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(),scores.std()*2))

[1.         0.91666667 0.91666667 0.90909091 0.9        0.88888889
 1.         1.         1.         1.        ]
10CV Accuracy: 0.95 (+/- 0.09)


C:\Users\user\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:605: Warning: The least populated class in y has only 4 members, which is too few. The minimum number of members in any class cannot be less than n_splits=10.
  % (min_groups, self.n_splits)), Warning)


### 5 ) . 5

##### RandomForest Classifier 

In [26]:
from sklearn.ensemble import RandomForestClassifier
clf=RandomForestClassifier(n_estimators=100)
clf.fit(X,Y) # Model
from sklearn.model_selection import cross_val_score
scores = cross_val_score(clf, X, Y, cv=10) # 10-fold cross validation
print(scores) # Results for all the folds
print("10CV Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(),scores.std()*2))
X_train , X_test , Y_train , Y_test = train_test_split(X, Y, test_size =0.333,random_state =0)
clf.fit(X_train ,Y_train) # clf is a classifier.
Y_pred = clf.fit(X_train , Y_train ).predict(X_test)
print(confusion_matrix(Y_test , Y_pred ))

C:\Users\user\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:605: Warning: The least populated class in y has only 4 members, which is too few. The minimum number of members in any class cannot be less than n_splits=10.
  % (min_groups, self.n_splits)), Warning)


[1.         0.91666667 0.91666667 0.90909091 1.         0.88888889
 1.         1.         1.         1.        ]
10CV Accuracy: 0.96 (+/- 0.09)
[[ 1  0  0  0  0  0  0]
 [ 0  6  0  0  0  0  0]
 [ 0  0  7  0  0  0  0]
 [ 0  0  0  2  0  0  0]
 [ 0  0  0  0  2  0  0]
 [ 0  0  0  0  0 12  0]
 [ 3  0  1  0  0  0  0]]


##### it has good accuracy like 0.96 compared with other classifications. KNeighborsClassifier has the best accuracy.